electrum/plugins/bitbox02/bitbox02.py

```diff
@@ -0,0 +1,628 @@
+
+    def show_address(
+        self, bip32_path: str, address_type: str, wallet: Deterministic_Wallet
+    ) -> str:
+
+        if self.bitbox02_device is None:
+            raise BaseException(
+                "Need to setup communication first before attempting any BitBox02 calls"
+            )
+
+        address_keypath = bip32.convert_bip32_path_to_list_of_uint32(bip32_path)
+        coin_network = self.coin_network_from_bip32_list(address_keypath)
+
+        if address_type == "p2wpkh":
+            script_config = bitbox02.btc.BTCScriptConfig(
+                simple_type=bitbox02.btc.BTCScriptConfig.P2WPKH
+            )
+        elif address_type == "p2wpkh-p2sh":
+            script_config = bitbox02.btc.BTCScriptConfig(
+                simple_type=bitbox02.btc.BTCScriptConfig.P2WPKH_P2SH
+            )
+        elif address_type == "p2wsh":
+            if type(wallet) is Multisig_Wallet:
+                script_config = self.btc_multisig_config(
+                    coin_network, address_keypath, wallet
+                )
+            else:
+                raise BaseException("Can only use p2wsh with multisig wallets")
+        else:
+            raise Exception(
+                "invalid address xtype: {} is not supported by the BitBox02".format(
+                    address_type
+                )
+            )
+
+        return self.bitbox02_device.btc_address(
+            keypath=address_keypath,
+            coin=coin_network,
+            script_config=script_config,
+            display=True,
+        )
+
+    def sign_transaction(
+        self,
+        keystore: Hardware_KeyStore,
+        tx: PartialTransaction,
+        wallet: Deterministic_Wallet,
+    ):
+        if tx.is_complete():
+            return
+
+        if self.bitbox02_device is None:
+            raise BaseException(
+                "Need to setup communication first before attempting any BitBox02 calls"
+            )
+
+        coin = bitbox02.btc.BTC
+        if constants.net.TESTNET:
+            coin = bitbox02.btc.TBTC
+
+        tx_script_type = None
+
+        # Build BTCInputType list
+        inputs = []
+        for txin in tx.inputs():
+            _, full_path = keystore.find_my_pubkey_in_txinout(txin)
+
+            inputs.append(
+                {
+                    "prev_out_hash": txin.prevout.txid[::-1],
+                    "prev_out_index": txin.prevout.out_idx,
+                    "prev_out_value": txin.value_sats(),
+                    "sequence": txin.nsequence,
+                    "keypath": full_path,
+                }
+            )
+
+            if tx_script_type == None:
+                tx_script_type = txin.script_type
+            elif tx_script_type != txin.script_type:
+                raise Exception("Cannot mix different input script types")
+        if full_path is None:
+            raise BaseException(
+                "A wallet owned pubkey was not found in the transaction input to be signed"
```

Comment:
is it okay that this test is outside the loop and hence will only test the last txin?

Reply:
Come to think of it, is there a scenario where inputs that we cannot sign are combined with inputs that we own in the electrum wallet?

Reply:
Not until recently. On master there is a "Combine" button in the advanced tx preview dialog, which allows constructing manual coinjoins. It hasn't really been tested with hw wallets though. Still, the answer is yes.

Reply:
Thanks for the clarification, I moved the check into the if clause now.